### ISSUE TITLE
"nDPI support for fragmented traffic not implemented" / No application recognition

### ISSUE THREAD

**Issue by tbard1** [pmacct/pmacct#428]
**Description**
pmacct with ndpi support, with both ipv4 and ipv6 traffic, no application recognition in elastiflow :

<img width="855" alt="image" src="https://user-images.githubusercontent.com/3597258/94614470-84286a80-02a6-11eb-88c4-6202e37e48a9.png">

then strange logs :

/var/log/messages 

```
Sep 29 22:46:23 xxxx pmacctd[15546]: WARN ( default/core ): nDPI support for fragmented traffic not implemented. If you see this message, please get in touch: Paolo Lucente <paolo@pmacct.net>
Sep 29 22:46:23 xxxx pmacctd[15546]: WARN ( default/core ): nDPI support for fragmented traffic not implemented. If you see this message, please get in touch: Paolo Lucente <paolo@pmacct.net>
Sep 29 22:46:23 xxxx pmacctd[15546]: WARN ( default/core ): nDPI support for fragmented traffic not implemented. If you see this message, please get in touch: Paolo Lucente <paolo@pmacct.net>
Sep 29 22:46:23 xxxx pmacctd[15546]: WARN ( default/core ): nDPI support for fragmented traffic not implemented. If you see this message, please get in touch: Paolo Lucente <paolo@pmacct.net>
Sep 29 22:46:24 xxxx pmacctd[15546]: INFO ( default/core ): NULL IPv6 payload length. Jumbo packets are currently not supported.
Sep 29 22:46:24 xxxx pmacctd[15546]: INFO ( default/core ): NULL IPv6 payload length. Jumbo packets are currently not supported.
```
/var/log/syslog

```
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): Building NetFlow v9 packet: offset = 48, template ID = 4097, total len = 28, # elements = 1
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): Building NetFlow v9 packet: offset = 72, template ID = 4097, total len = 52, # elements = 2
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): Sending NetFlow v9/IPFIX packet: len = 72
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): Sent 144 netflow packets
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): EXPIRED: seq:6497 [A.A.A.A]:443 <> [192.168.B.B]:46846 proto:6 octets>:187827 packets>:105 octets<:2313 packets<:33 start:2020-09-29T22:50:00.889 finish:2020-09-29T22:50:01.915 tcp>:1b tcp<:1b flowlabel>:00000000 flowlabel<:00000000  (0x55e43bbd2040)
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): EXPIRED: seq:6499 [192.168.C.C]:56122 <> [192.168.D.D]:42300 proto:6 octets>:679 packets>:8 octets<:2412 packets<:6 start:2020-09-29T22:50:01.664 finish:2020-09-29T22:50:01.674 tcp>:1b tcp<:1b flowlabel>:00000000 flowlabel<:00000000  (0x55e43bb66ad0)
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): EXPIRED: seq:6500 [192.168.E.E]:56123 <> [192.168.F.F]:42300 proto:6 octets>:1048 packets>:15 octets<:13505 packets<:9 start:2020-09-29T22:50:01.674 finish:2020-09-29T22:50:01.684 tcp>:1b tcp<:1b flowlabel>:00000000 flowlabel<:00000000  (0x55e43bbd1ef0)
Sep 29 22:56:01 xxxx pmacctd[15547]: DEBUG ( default_nfprobe/nfprobe ): EXPIRED: seq:6501 [192.168.G.G]:56124 <> [192.168.H.H]:42300 proto:6 octets>:742 packets>:9 octets<:3894 packets<:6 start:2020-09-29T22:50:01.687 finish:2020-09-29T22:50:01.697 tcp>:1b tcp<:1b flowlabel>:00000000 flowlabel<:00000000  (0x55e43bb18910)
```

**Version**
pmacct 1.7.6
ndpi 3.2 stable
Debian GNU/Linux 10 \n \l

Any help appreciated ! so close !

**Comment by paololucente** [comment]
Hi Thomas,

If the strange logs are occasional, you can forget about them, at least in a first pass.

I am not familiar with your collector so it is possible that there is a compatibility issue; pmacct exports application in the form of an ID and such ID should be mapped to a name at the collector side (collector being also compiled against nDPI and having knowledge of such IDs). I think in fact in your graph you are reported the distribution of the application IDs rather than application names.

Paolo

**Comment by tbard1** [comment]
Roger that. 
I'd have a look toward ElastiFlow.

Thanks a lots. 

Thomas

**Comment by paololucente** [comment]
Hi Thomas ( @tbard1 ),

I took a moment to test everything and just passed a minor fix ( https://github.com/pmacct/pmacct/commit/463d22f82181524041f6a288808368a055eb3cf3 ). I may have been a tad mis-leading with my previous comment: pmacct exports flows with an application ID - which i guess is the number you see mentioned in your report. On top of that, a bit like a Cisco router does with NBAR, it exports Option Data packets where application ID maps to an application Name (this is the part i missed in my previous comment).

In summary: the collector should be able to parse the Option packets to get ID -> Name mappings and, when a flow gets in with application ID X, it knows ho to translate that into Application Name Y. 

Hope this helps. I'd go ahead and close this issue.

Paolo

**Comment by tbard1** [comment]
Hi Paolo, 

Thank you for your time on my case

Unfortunately, no success. I re-compiled pmacct with the fix, but same result : 

<img width="593" alt="Capture d’écran 2020-10-07 à 22 37 31" src="https://user-images.githubusercontent.com/3597258/95385573-79955300-08ee-11eb-8900-aec1b1628ad3.png">

FYI here is my pmacct.conf : 

`daemonize: true`
`#daemonize: false`
`interface: enp1s0`
`aggregate: src_host, dst_host, src_port, dst_port, proto, tos, class`
`plugins: nfprobe`
`plugin_pipe_size: 10240000`
`plugin_buffer_size: 10240`
`nfprobe_receiver: 192.168.0.1:2055`
`nfprobe_version: 9`
`!nfprobe_timeouts: tcp=30:maxlife=60`

**Comment by paololucente** [comment]
Hi Thomas ( @tbard1 ),

Sure, this is something the collector you are using should understand & implement. pmacct simply performs similarly to how a Cisco router would do with NBAR - as i was saying in my previous message: it gives the collector all needed info, the collector should take the pieces and combine them. Closing the issue for now.

Paolo
